crates/roc_host/src/roc.rs

```diff
@@ -771,170 +785,313 @@ fn os_str_to_roc_path(os_str: &std::ffi::OsStr) -> RocList<u8> {
-    query: &roc_std::RocStr,
-    bindings: &roc_std::RocList<SQLiteBindings>,
-) -> roc_std::RocResult<RocList<RocList<roc_app::SQLiteValue>>, SQLiteError> {
+fn thread_local_prepare(
+    stmt: &SqliteStatement,
+) -> Result<*mut sqlite3_sys::sqlite3_stmt, SqliteError> {
     // Get the connection
     let connection = {
-        match get_connection(db_path.as_str()) {
-            Ok(c) => c,
-            Err(err) => {
-                return RocResult::err(SQLiteError {
-                    code: err.code.unwrap_or_default() as i64,
-                    message: RocStr::from(err.message.unwrap_or_default().as_str()),
-                });
-            }
+        match get_connection(&stmt.db_path.as_str()) {
+            Ok(conn) => conn,
+            Err(err) => return Err(err),
         }
     };
 
-    // Prepare the query
-    let mut statement = {
-        match connection.prepare(query.as_str()) {
-            Ok(c) => c,
-            Err(err) => {
-                return RocResult::err(SQLiteError {
-                    code: err.code.unwrap_or_default() as i64,
-                    message: RocStr::from(err.message.unwrap_or_default().as_str()),
-                });
+    stmt.stmt
+        .get_or_try(|| {
+            let mut unsafe_stmt = UnsafeStmt(std::ptr::null_mut());
+            let err = unsafe {
+                sqlite3_sys::sqlite3_prepare_v2(
+                    connection,
+                    stmt.query.as_str().as_ptr() as *const c_char,
+                    stmt.query.len() as i32,
+                    &mut unsafe_stmt.0,
+                    std::ptr::null_mut(),
+                )
+            };
+            if err != sqlite3_sys::SQLITE_OK {
+                return Err(err);
             }
-        }
+            return Ok(unsafe_stmt);
+        })
+        .map(|x| x.0)
+}
+
+#[roc_fn(name = "sqlitePrepare")]
+fn sqlite_prepare(
+    db_path: &roc_std::RocStr,
+    query: &roc_std::RocStr,
+) -> roc_std::RocResult<RocBox<()>, roc_app::SqliteError> {
+    // Prepare the query
+    let stmt = SqliteStatement {
+        db_path: db_path.clone(),
+        query: query.clone(),
+        stmt: ThreadLocal::new(),
     };
 
-    // Add bindings for the query
-    for binding in bindings {
-        match statement.bind(binding) {
-            Ok(()) => {}
-            Err(err) => {
-                return RocResult::err(SQLiteError {
-                    code: err.code.unwrap_or_default() as i64,
-                    message: RocStr::from(err.message.unwrap_or_default().as_str()),
-                });
-            }
+    // Always prepare once to ensure no errors and prep for current thread.
+    if let Err(err) = thread_local_prepare(&stmt) {
+        return roc_err_from_sqlite_err(err);
+    }
+
+    let heap = sqlite_stmt_heap();
+    let alloc_result = heap.alloc_for(stmt);
+    match alloc_result {
+        Ok(out) => {
+            // TODO: only make these constant if they escape init.
```

Comment:
I do think this TODO needs to be dealt with before we land. Just an FYI

Reply:
@bhansconnect  I think I resolved this in fd317fb. Can you just double check as I wasn't 100% around the logic here. Like are we setting refcount to zero during or after init. I've opted for during as that feels right, these stmts will escape init and therefore we don't want them to be freed.

Reply:
I think I realized a cooler way to do this. Just pushed it. It also has the advantage that an sqlite statement that is freed during init will actually get freed. Only those still live after init will be promoted to constants!